src/BridgeFactory.sol

```diff
@@ -111,29 +120,57 @@ contract BridgeFactory is IERC1271, MultiCallContext, TwoStepOwnable {
                 revert(0x1c, 0x04)
             }
 
-            // set owner in proxy
+            // set the pending owner in the proxy
             mstore(0x14, owner)
             mstore(0x00, 0xc42069ec000000000000000000000000) // selector for `setPendingOwner(address)` with padding for owner
-            pop(call(gas(), proxy, 0x00, 0x10, 0x24, 0x00, 0x00))
+            if iszero(call(gas(), proxy, 0x00, 0x10, 0x24, 0x00, 0x00)) {
+                returndatacopy(ptr, 0x00, returndatasize())
+                revert(ptr, returndatasize())
+            }
+
+            // restore clobbered memory
+            mstore(0x40, ptr)
         }
     }
 
     function setPendingOwner(address owner) external onlyFactory {
         _setPendingOwner(owner);
     }
 
-    function approvePermit2(IERC20 token) external onlyProxy returns (bool) {
-        token.safeApprove(0x000000000022D473030F116dDEE9F6B43aC78BA3, type(uint256).max);
+    function approvePermit2(IERC20 token, uint256 amount) external onlyProxy returns (bool) {
+        token.safeApprove(0x000000000022D473030F116dDEE9F6B43aC78BA3, amount);
         return true;
     }
 
-    function call(address target, uint256 value, bytes calldata data) external onlyOwner returns (bytes memory) {
+    function call(address payable target, uint256 value, bytes calldata data)
+        external
+        onlyOwner
+        returns (bytes memory)
+    {
         (bool success, bytes memory result) = target.call{value: value}(data);
-        require(success);
-        return result;
+        success.maybeRevert(result);
+        assembly ("memory-safe") {
+            let start := sub(result, 0x20) // TODO: examine bytecode to ensure this does not clobber reserved memory
+            mstore(start, 0x20)
+            return(start, add(0x40, mload(result)))
+        }
     }
 
-    function cleanup() external onlyOwner {
-        selfdestruct(payable(_msgSender()));
+    function cleanup(address payable beneficiary) external onlyProxy {
+        address owner_ = owner();
+        if (_msgSender() != owner_) {
+            if (owner_ != address(0)) {
+                assembly ("memory-safe") {
+                    mstore(0x00, 0x1e092104) // selector for `PermissionDenied()`
+                    revert(0x1c, 0x04)
+                }
+            }
+            address pendingOwner_ = pendingOwner();
+            if (pendingOwner_ == address(0)) {
+                selfdestruct(beneficiary);
+            }
```

Comment:
@e1Ru1o what do you think of this clause? I'm kinda torn on it. This effectively means that on pre-Cancun chains, calling `renounceOwnership` really just resets the owner to the initial owner. I have less of a feeling about the fact that this means that the ETH in a renounced proxy can be taken by anybody, because that ETH was already basically lost.

Reply:
I see you changed it now. I would let only the owner call cleanup, if the `pendingOwner` wants the cleanup to be done he needs to first `acceptOwnership`. This is in favor of keeping the logic clean/simple.

I see as well the logic for calling cleanup from `deploy`, not a fan at the moment, it seems to me it is for rescuing ETH that was sent to proxy after it was already cleaned up, is that so? if so, it is a very intricate case, I would not expect assets being sent there anymore so if it happens I will let the user do the rescue. You can have a separated `rescue` function for that if you want that does the `deploy` + `cleanup`, but even without it a user can do it.